by-language/csharp-npgsql/DemoTypes.cs

```diff
@@ -0,0 +1,188 @@
+        text TEXT,
+        char CHARACTER(5),
+        timestamp_tz TIMESTAMP WITH TIME ZONE,
+        timestamp_notz TIMESTAMP WITHOUT TIME ZONE,
+        ip IP,
+        -- Container types
+        ""array"" ARRAY(STRING),
+        ""object"" OBJECT(DYNAMIC),
+        -- Geospatial types
+        geopoint GEO_POINT,
+        geoshape GEO_SHAPE,
+        -- Vector type
+        ""float_vector"" FLOAT_VECTOR(3)
+    );
+            ", conn))
+            {
+                cmd.ExecuteNonQuery();
+            }
+
+            // Insert single data point.
+            await using (var cmd = new NpgsqlCommand(@"
+    INSERT INTO testdrive.example (
+        null_integer,
+        integer,
+        bigint,
+        float,
+        double,
+        decimal,
+        bit,
+        bool,
+        text,
+        char,
+        timestamp_tz,
+        timestamp_notz,
+        ip,
+        ""array"",
+        -- ""object"",
+        geopoint,
+        -- geoshape,
+        float_vector
+    ) VALUES (
+        @null_integer,
+        @integer,
+        @bigint,
+        @float,
+        @double,
+        @decimal,
+        @bit,
+        @bool,
+        @text,
+        @char,
+        @timestamp_tz,
+        @timestamp_notz,
+        @ip,
+        @array,
+        -- @object,
+        @geopoint,
+        -- @egoshape,
+        @float_vector
+    );
+", conn))
+            {
+                Console.WriteLine(cmd);
+                cmd.Parameters.AddWithValue("null_integer", DBNull.Value);
+                cmd.Parameters.AddWithValue("integer", 42);
+                cmd.Parameters.AddWithValue("bigint", 42);
+                cmd.Parameters.AddWithValue("float", 42.42);
+                cmd.Parameters.AddWithValue("double", 42.42);
+                cmd.Parameters.AddWithValue("decimal", 42.42);
+                cmd.Parameters.AddWithValue("bit", "01010101");
+                cmd.Parameters.AddWithValue("bool", true);
+                cmd.Parameters.AddWithValue("text", "foobar");
+                cmd.Parameters.AddWithValue("char", "foo");
+                cmd.Parameters.AddWithValue("timestamp_tz", "1970-01-02T00:00:00+01:00");
+                cmd.Parameters.AddWithValue("timestamp_notz", "1970-01-02T00:00:00");
+                cmd.Parameters.AddWithValue("ip", "127.0.0.1");
+                cmd.Parameters.AddWithValue("array", new List<string>{"foo", "bar"});
+                // FIXME: System.NotSupportedException: Cannot resolve 'hstore' to a fully qualified datatype name. The datatype was not found in the current database info.
+                // cmd.Parameters.AddWithValue("object", new Dictionary<string, string>(){{"foo", "bar"}});
+                cmd.Parameters.AddWithValue("geopoint", new List<double>{85.43, 66.23});
+                // FIXME: Npgsql.PostgresException : XX000: line 38:9: no viable alternative at input 'VALUES
+                // cmd.Parameters.AddWithValue("geoshape", "POLYGON ((5 5, 10 5, 10 10, 5 10, 5 5))");
```

Comment:
Maybe GEOSHAPE can be made work easily. I haven't yet investigated what is behind a hiccup in this case. The primary reason of this patch is to evaluate and educate, and to set the stage for relevant insights and discussions.

Reply:
Basic support for GEO_SHAPE types (marshalled using STRING) is now validated per 131972aef16.